File: sdk/java/src/main/java/com/rajatim/zhtw/JsonAdapter.java
package com.rajatim.zhtw;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.function.Function;
import java.util.regex.Pattern;

/** Exact-byte JSON string-value conversion. */
final class JsonAdapter {
    private static final Pattern NUMBER = Pattern.compile(
            "-?(?:0|[1-9][0-9]*)(?:\\.[0-9]+)?(?:[eE][+-]?[0-9]+)?");

    private JsonAdapter() {}

    static String convert(String text, Function<String, String> converter) {
        if (text == null) return null;
        Scanner originalScanner = new Scanner(text);
        Node original = originalScanner.parse();
        List<Replacement> replacements = new ArrayList<>();
        for (StringToken token : originalScanner.values) {
            String target = converter.apply(token.value);
            if (target == null) {
                throw new JsonAdapterException(
                        "JSON value converter must return a string",
                        "invalid_converter_result");
            }
            if (!target.equals(token.value)) {
                replacements.add(new Replacement(token.start, token.end, quote(target)));
            }
        }
        if (replacements.isEmpty()) return text;

        StringBuilder output = new StringBuilder(text.length());
        int lastEnd = 0;
        for (Replacement replacement : replacements) {
            output.append(text, lastEnd, replacement.start).append(replacement.value);
            lastEnd = replacement.end;
        }
        output.append(text, lastEnd, text.length());
        String convertedText = output.toString();
        Node converted = new Scanner(convertedText).parse();
        if (!original.equals(converted)) {
            throw new JsonAdapterException(
                    "converted JSON changed non-value structure", "structure_changed");
        }
        return convertedText;
    }

    static String quote(String value) {
        validateUnicode(value);
        StringBuilder result = new StringBuilder(value.length() + 2).append('"');
        for (int i = 0; i < value.length(); ) {
            int cp = value.codePointAt(i);
            switch (cp) {
                case '"': result.append("\\\""); break;
                case '\\': result.append("\\\\"); break;
                case '\b': result.append("\\b"); break;
                case '\f': result.append("\\f"); break;
                case '\n': result.append("\\n"); break;
                case '\r': result.append("\\r"); break;
                case '\t': result.append("\\t"); break;
                default:
                    if (cp <= 0x1f) result.append(String.format("\\u%04x", cp));
                    else result.appendCodePoint(cp);
            }
            i += Character.charCount(cp);
        }
        return result.append('"').toString();
    }

    private static void validateUnicode(String value) {
        for (int i = 0; i < value.length(); i++) {
            char current = value.charAt(i);
            if (Character.isHighSurrogate(current)) {
                if (i + 1 >= value.length() || !Character.isLowSurrogate(value.charAt(i + 1))) {
                    throw invalid("JSON contains an unpaired Unicode surrogate");
                }
                i++;
            } else if (Character.isLowSurrogate(current)) {
                throw invalid("JSON contains an unpaired Unicode surrogate");
            }
        }
    }

    private static JsonAdapterException invalid(String message) {
        return new JsonAdapterException(message, "invalid_json");
    }

    private static final class Replacement {
        final int start;
        final int end;
        final String value;

        Replacement(int start, int end, String value) {
            this.start = start;
            this.end = end;
            this.value = value;
        }
    }

    private static final class StringToken {
        final int start;
        final int end;
        final String value;

        StringToken(int start, int end, String value) {
            this.start = start;
            this.end = end;
            this.value = value;
        }
    }

    private static final class Node {
        final String kind;
        final String value;
        final List<Node> children;

        Node(String kind, String value, List<Node> children) {
            this.kind = kind;
            this.value = value;
            this.children = children;
        }

        @Override
        public boolean equals(Object other) {
            if (this == other) return true;
            if (!(other instanceof Node)) return false;
            Node node = (Node) other;
            return Objects.equals(kind, node.kind)
                    && Objects.equals(value, node.value)
                    && Objects.equals(children, node.children);
        }

        @Override
        public int hashCode() { return Objects.hash(kind, value, children); }
    }

    private static final class Scanner {
        final String text;
        final List<StringToken> values = new ArrayList<>();
        int position;

        Scanner(String text) { this.text = text; }

        Node parse() {
            skipWhitespace();
            Node result = parseValue(true);
            skipWhitespace();
            if (position != text.length()) throw invalid("trailing JSON content");
            return result;
        }

        private Node parseValue(boolean collectString) {
            skipWhitespace();
            if (position >= text.length()) throw invalid("missing JSON value");
            char current = text.charAt(position);
            if (current == '"') {
                parseString(collectString);
                return new Node("string", null, List.of());
            }
            if (current == '{') return parseObject();
            if (current == '[') return parseArray();
            int start = position;
            while (position < text.length()
                    && ",]} \t\r\n".indexOf(text.charAt(position)) < 0) position++;
            String primitive = text.substring(start, position);
            if (!(primitive.equals("true") || primitive.equals("false")
                    || primitive.equals("null") || NUMBER.matcher(primitive).matches())) {
                throw invalid("invalid JSON primitive");
            }
            return new Node("primitive", primitive, List.of());
        }

        private Node parseObject() {
            expect('{');
            skipWhitespace();
            List<Node> children = new ArrayList<>();
            Set<String> keys = new HashSet<>();
            if (peek('}')) {
                position++;
                return new Node("object", null, children);
            }
            while (true) {
                skipWhitespace();
                if (!peek('"')) throw invalid("JSON object key must be a string");
                String key = parseString(false);
                if (!keys.add(key)) {
                    throw new JsonAdapterException(
                            "duplicate JSON object key", "duplicate_key");
                }
                skipWhitespace();
                expect(':');
                children.add(new Node("member", key, List.of(parseValue(true))));
                skipWhitespace();
                if (peek('}')) {
                    position++;
                    return new Node("object", null, children);
                }
                expect(',');
            }
        }

        private Node parseArray() {
            expect('[');
            skipWhitespace();
            List<Node> children = new ArrayList<>();
            if (peek(']')) {
                position++;
                return new Node("array", null, children);
            }
            while (true) {
                children.add(parseValue(true));
                skipWhitespace();
                if (peek(']')) {
                    position++;
                    return new Node("array", null, children);
                }
                expect(',');
            }
        }

        private String parseString(boolean collect) {
            int start = position;
            expect('"');
            StringBuilder decoded = new StringBuilder();
            while (position < text.length()) {
                char current = text.charAt(position++);
                if (current == '"') {
                    String value = decoded.toString();
                    validateUnicode(value);
                    if (collect) values.add(new StringToken(start, position, value));
                    return value;
                }
                if (current < 0x20) throw invalid("unescaped JSON control character");
                if (current != '\\') {
                    decoded.append(current);
                    continue;
                }
                if (position >= text.length()) throw invalid("unterminated JSON escape");
                char escaped = text.charAt(position++);
                switch (escaped) {
                    case '"': case '\\': case '/': decoded.append(escaped); break;
                    case 'b': decoded.append('\b'); break;
                    case 'f': decoded.append('\f'); break;
                    case 'n': decoded.append('\n'); break;
                    case 'r': decoded.append('\r'); break;
                    case 't': decoded.append('\t'); break;
                    case 'u':
                        if (position + 4 > text.length()) throw invalid("short Unicode escape");
                        int value = 0;
                        for (int i = 0; i < 4; i++) {
                            int digit = Character.digit(text.charAt(position++), 16);
                            if (digit < 0) throw invalid("invalid Unicode escape");
                            value = value * 16 + digit;
                        }
                        decoded.append((char) value);
                        break;
                    default: throw invalid("invalid JSON escape");
                }
            }
            throw invalid("unterminated JSON string");
        }

        private void skipWhitespace() {
            while (position < text.length()
                    && " \t\r\n".indexOf(text.charAt(position)) >= 0) position++;
        }

        private boolean peek(char expected) {
            return position < text.length() && text.charAt(position) == expected;
        }

        private void expect(char expected) {
            if (!peek(expected)) throw invalid("unexpected JSON token");
            position++;
        }
    }
}

File: sdk/dotnet/src/JsonAdapter.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace Zhtw
{
    public sealed class JsonAdapterException : Exception
    {
        public string Code { get; }

        internal JsonAdapterException(string code, string message)
            : base("zhtw JSON adapter " + code + ": " + message)
        {
            Code = code;
        }
    }

    internal static class JsonAdapter
    {
        private sealed class Node
        {
            internal char Kind;
            internal string Primitive;
            internal List<Node> Array;
            internal List<Member> Object;

            internal bool SameStructure(Node other)
            {
                if (other == null || Kind != other.Kind || Primitive != other.Primitive) return false;
                if (Array != null)
                {
                    if (other.Array == null || Array.Count != other.Array.Count) return false;
                    for (int i = 0; i < Array.Count; i++)
                        if (!Array[i].SameStructure(other.Array[i])) return false;
                }
                if (Object != null)
                {
                    if (other.Object == null || Object.Count != other.Object.Count) return false;
                    for (int i = 0; i < Object.Count; i++)
                    {
                        if (Object[i].Key != other.Object[i].Key ||
                            !Object[i].Value.SameStructure(other.Object[i].Value)) return false;
                    }
                }
                return true;
            }
        }

        private sealed class Member
        {
            internal string Key;
            internal Node Value;
        }

        private sealed class StringToken
        {
            internal int Start;
            internal int End;
            internal string Value;
        }

        private sealed class Scanner
        {
            private readonly string _text;
            private int _position;
            internal List<StringToken> Values { get; } = new List<StringToken>();

            internal Scanner(string text)
            {
                _text = text ?? throw new ArgumentNullException(nameof(text));
            }

            internal Node Parse()
            {
                SkipWhitespace();
                Node node = ParseValue(true);
                SkipWhitespace();
                if (_position != _text.Length) throw Invalid("trailing JSON content");
                return node;
            }

            private Node ParseValue(bool collectString)
            {
                SkipWhitespace();
                char current = Peek();
                if (current == '"')
                {
                    ParseString(collectString);
                    return new Node { Kind = 's' };
                }
                if (current == '{') return ParseObject();
                if (current == '[') return ParseArray();
                if (current == '\0') throw Invalid("missing JSON value");
                int start = _position;
                while (_position < _text.Length &&
                    ",]} \t\r\n".IndexOf(_text[_position]) < 0) _position++;
                string primitive = _text.Substring(start, _position - start);
                if (primitive != "true" && primitive != "false" && primitive != "null" &&
                    !ValidNumber(primitive)) throw Invalid("invalid JSON primitive");
                return new Node { Kind = 'p', Primitive = primitive };
            }

            private Node ParseObject()
            {
                Expect('{');
                SkipWhitespace();
                var members = new List<Member>();
                var keys = new HashSet<string>(StringComparer.Ordinal);
                if (Peek() == '}')
                {
                    _position++;
                    return new Node { Kind = 'o', Object = members };
                }
                while (true)
                {
                    SkipWhitespace();
                    if (Peek() != '"') throw Invalid("JSON object key must be a string");
                    string key = ParseString(false);
                    if (!keys.Add(key)) throw Error("duplicate_key", "duplicate JSON object key");
                    SkipWhitespace();
                    Expect(':');
                    members.Add(new Member { Key = key, Value = ParseValue(true) });
                    SkipWhitespace();
                    if (Peek() == '}')
                    {
                        _position++;
                        return new Node { Kind = 'o', Object = members };
                    }
                    Expect(',');
                }
            }

            private Node ParseArray()
            {
                Expect('[');
                SkipWhitespace();
                var values = new List<Node>();
                if (Peek() == ']')
                {
                    _position++;
                    return new Node { Kind = 'a', Array = values };
                }
                while (true)
                {
                    values.Add(ParseValue(true));
                    SkipWhitespace();
                    if (Peek() == ']')
                    {
                        _position++;
                        return new Node { Kind = 'a', Array = values };
                    }
                    Expect(',');
                }
            }

            private string ParseString(bool collect)
            {
                int start = _position;
                Expect('"');
                var decoded = new StringBuilder();
                while (_position < _text.Length)
                {
                    char current = _text[_position++];
                    if (current == '"')
                    {
                        string value = decoded.ToString();
                        if (collect) Values.Add(new StringToken { Start = start, End = _position, Value = value });
                        return value;
                    }
                    if (current < 0x20) throw Invalid("unescaped JSON control character");
                    if (current != '\\')
                    {
                        if (char.IsHighSurrogate(current))
                        {
                            if (_position >= _text.Length || !char.IsLowSurrogate(_text[_position]))
                                throw Invalid("unpaired Unicode surrogate");
                            decoded.Append(current);
                            decoded.Append(_text[_position++]);
                        }
                        else
                        {
                            if (char.IsLowSurrogate(current)) throw Invalid("unpaired Unicode surrogate");
                            decoded.Append(current);
                        }
                        continue;
                    }
                    char escaped = Next();
                    switch (escaped)
                    {
                        case '"': decoded.Append('"'); break;
                        case '\\': decoded.Append('\\'); break;
                        case '/': decoded.Append('/'); break;
                        case 'b': decoded.Append('\b'); break;
                        case 'f': decoded.Append('\f'); break;
                        case 'n': decoded.Append('\n'); break;
                        case 'r': decoded.Append('\r'); break;
                        case 't': decoded.Append('\t'); break;
                        case 'u':
                            char first = (char)ParseHexQuad();
                            if (char.IsHighSurrogate(first))
                            {
                                if (Next() != '\\' || Next() != 'u')
                                    throw Invalid("unpaired Unicode surrogate");
                                char second = (char)ParseHexQuad();
                                if (!char.IsLowSurrogate(second))
                                    throw Invalid("unpaired Unicode surrogate");
                                decoded.Append(first).Append(second);
                            }
                            else
                            {
                                if (char.IsLowSurrogate(first)) throw Invalid("unpaired Unicode surrogate");
                                decoded.Append(first);
                            }
                            break;
                        default: throw Invalid("invalid JSON escape");
                    }
                }
                throw Invalid("unterminated JSON string");
            }

            private int ParseHexQuad()
            {
                int value = 0;
                for (int i = 0; i < 4; i++)
                {
                    char current = Next();
                    int digit;
                    if (current >= '0' && current <= '9') digit = current - '0';
                    else if (current >= 'a' && current <= 'f') digit = current - 'a' + 10;
                    else if (current >= 'A' && current <= 'F') digit = current - 'A' + 10;
                    else throw Invalid("invalid Unicode escape");
                    value = value * 16 + digit;
                }
                return value;
            }

            private void SkipWhitespace()
            {
                while (_position < _text.Length && " \t\r\n".IndexOf(_text[_position]) >= 0)
                    _position++;
            }

            private char Peek() => _position < _text.Length ? _text[_position] : '\0';

            private char Next()
            {
                if (_position >= _text.Length) throw Invalid("unexpected end of JSON");
                return _text[_position++];
            }

            private void Expect(char expected)
            {
                if (Next() != expected) throw Invalid("unexpected JSON token");
            }
        }

        internal static string ConvertValues(Converter converter, string text)
        {
            var originalScanner = new Scanner(text);
            Node original = originalScanner.Parse();
            var output = new StringBuilder(text.Length);
            int lastEnd = 0;
            bool changed = false;
            foreach (var token in originalScanner.Values)
            {
                string target = converter.Convert(token.Value);
                if (target == token.Value) continue;
                output.Append(text, lastEnd, token.Start - lastEnd);
                output.Append(QuoteString(target));
                lastEnd = token.End;
                changed = true;
            }
            if (!changed) return text;
            output.Append(text, lastEnd, text.Length - lastEnd);
            string convertedText = output.ToString();
            Node converted = new Scanner(convertedText).Parse();
            if (!original.SameStructure(converted))
                throw Error("structure_changed", "converted JSON changed non-value structure");
            return convertedText;
        }

        internal static string QuoteString(string value)
        {
            var output = new StringBuilder(value.Length + 2);
            output.Append('"');
            for (int i = 0; i < value.Length; i++)
            {
                char current = value[i];
                switch (current)
                {
                    case '"': output.Append("\\\""); break;
                    case '\\': output.Append("\\\\"); break;
                    case '\b': output.Append("\\b"); break;
                    case '\f': output.Append("\\f"); break;
                    case '\n': output.Append("\\n"); break;
                    case '\r': output.Append("\\r"); break;
                    case '\t': output.Append("\\t"); break;
                    default:
                        if (current < 0x20)
                            output.Append("\\u").Append(((int)current).ToString("x4", CultureInfo.InvariantCulture));
                        else
                            output.Append(current);
                        break;
                }
            }
            output.Append('"');
            return output.ToString();
        }

        private static bool ValidNumber(string value)
        {
            int position = 0;
            if (position < value.Length && value[position] == '-') position++;
            if (position >= value.Length) return false;
            if (value[position] == '0') position++;
            else if (value[position] >= '1' && value[position] <= '9')
            {
                while (position < value.Length && IsAsciiDigit(value[position])) position++;
            }
            else return false;
            if (position < value.Length && value[position] == '.')
            {
                position++;
                int start = position;
                while (position < value.Length && IsAsciiDigit(value[position])) position++;
                if (position == start) return false;
            }
            if (position < value.Length && (value[position] == 'e' || value[position] == 'E'))
            {
                position++;
                if (position < value.Length && (value[position] == '+' || value[position] == '-')) position++;
                int start = position;
                while (position < value.Length && IsAsciiDigit(value[position])) position++;
                if (position == start) return false;
            }
            return position == value.Length;
        }

        private static bool IsAsciiDigit(char value) => value >= '0' && value <= '9';

        private static JsonAdapterException Invalid(string message) =>
            Error("invalid_json", message);

        private static JsonAdapterException Error(string code, string message) =>
            new JsonAdapterException(code, message);
    }
}
